### DISCUSSION TITLE
tutorials/scripting/gdscript/gdscript_exports

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#143] [User-contributed notes]
# tutorials/scripting/gdscript/gdscript_exports

In Godot, class members can be exported. This means their value gets saved along with the resource (such as the scene) they're attached to. They will also be available for editing in the property e...

https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_exports.html

**Comment by Lazy-Rabbit-2001** [comment]
What about `@export_custom`?

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/9993

**Comment by berarma** [comment]
There's no mention that a description text can be added to exported vars starting the preceding line with ##.

```gdscript
## This is my description text
@export var a: int
```

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10097

**Comment by RensTillmann** [comment]
What if you want to listen to a value being changed for exported properties, for instance a bool value changing from true to false. And preferably outside of `_process()`? Is it possible?

**Reply by Kartopod** [reply]
You can make a setter on the property. Note that you would need to make the script @tool for this. 
```gdscript
@tool

signal my_property_changed(new_value: bool)
@export var my_property: bool:
	set(value):
		my_property_changed.emit(value)
```
Also note that you may need to reload the scene/editor for this to take effect.

**Comment by dillydadally** [comment]
What about if you only want to specify a min or a max to a value instead of a range? For example, if I am exporting a variable called max_health, I want to set the min to 1 (as 0 for max health makes no sense), but I don't want to provide an upper limit. How would I do that? 

It seems like an @export_min and @export_max might be nice or a keyword you can use in the @export_range (note that INF won't work for ints I believe).

**Reply by Calinou** [reply]
If you specify `or_greater` and/or `or_less` at the end of the hint string, values won't be limited for the upper/lower bound. The upper/lower bound you specify will only be used for the slider display, not for numeric input or dragging the number (unless you drag the *slider* directly).

**Reply by dillydadally** [reply]
Great idea! Still not a perfect solution but a much better one that I thought existed at present!

**Comment by mr-ai-solutions** [comment]
Hello people, I get some script from 2022 version of Godot. There is a code write in this format:

'''export (int, 0, 256) var max_polyphony:int = 96 setget set_max_polyphony'''

Can you help me to parse it to actual 4.3 GDScript? Thanks.

**Reply by Coxcopi** [reply]
```gdscript
@export_range(0, 256) var max_polyphony : int = 96: set = set_max_polyphony
```

**Comment by GersonFeDutra** [comment]
How do I export min-max-values like the ones in the particle system?
![min-max-values in the particle system](https://i.imgur.com/e7FqEf5.png)

**Reply by Calinou** [reply]
This uses a dedicated system which can't be replicated in scripts or editor plugins yet: https://github.com/godotengine/godot/pull/81260#issuecomment-1866972511

**Comment by ImmaB** [comment]
I had the Problem that when I changed an exported property from a tool script the scene was not marked as "modified". The solution was to use undo_redo. Maybe this can be mentioned here

**Reply by Calinou** [reply]
I think this would be better mentioned in [Running code in the editor](https://docs.godotengine.org/en/stable/tutorials/plugins/running_code_in_the_editor.html).

**Comment by RudyFisher7** [comment]
```gdscript
@export 
```
requires a type so Variant can't be exported. Here is a partial, simple work-around.
```gdscript
class_name ExportableVariant
extends Resource


enum Type {
	BOOL = TYPE_BOOL,
	INT = TYPE_INT,
	FLOAT = TYPE_FLOAT,
	VECTOR2 = TYPE_VECTOR2,
	VECTOR3 = TYPE_VECTOR3,
}


@export var type: Type = Type.BOOL
@export var value_bool: bool = false
@export var value_int: int = 0
@export var value_float: float = 0.0
@export var value_vector_2 := Vector2()
@export var value_vector_3 := Vector3()


static func from_float(value: float) -> ExportableVariant:
	var result := ExportableVariant.new()
	
	result.type = Type.FLOAT
	result.value_float = value
	
	return result


func get_value():
	match type:
		Type.BOOL:
			return value_bool
		Type.INT:
			return value_int
		Type.FLOAT:
			return value_float
		Type.VECTOR2:
			return value_vector_2
		Type.VECTOR3:
			return value_vector_3
```

This can be useful when you want to tweak easily tweak animations between instances with property tweeners. For example, an openable container class (some containers you might want to use a hinge joint, others you might want to change color).
```gdscript
extends Openable


@export_group("Required Nodes")
@export var hinge: Node3D = null

@export_group("Animation Properties", "animation")
@export var animation_property_path: String = "rotation:x"
@export var animation_final_value := ExportableVariant.from_float(-150)
@export var animation_time: float = 2.5
@export var animation_ease_type: Tween.EaseType = Tween.EaseType.EASE_IN
@export var animation_transition_type: Tween.TransitionType = Tween.TransitionType.TRANS_CIRC

func _open() -> void:
	_is_tweening = true
	var tween: Tween = create_tween()
	tween.tween_property(hinge, animation_property_path, deg_to_rad(animation_final_value.get_value()), animation_time).set_ease(animation_ease_type).set_trans(animation_transition_type)
```

**Reply by Calinou** [reply]
> requires a type so Variant can't be exported. Here is a partial, simple work-around.

Note that there is a PR addressing this: https://github.com/godotengine/godot/pull/89324

**Comment by iRadEntertainment** [comment]
Regarding `@export_tool_button`, it would be nice to have mentioned where to find the list of the Engine Icons.
It could be a good idea to mention to have a look at the https://github.com/godotengine/godot/tree/master/editor/icons list in this doc page.

**Comment by KingTempest07** [comment]
### 1.
Will there ever be anything to read further? Every time I come here and read to the bottom where it says "before reading further" and kinda wish there was something else to read on like you'd expect lol. 

While I'm sure one could just figure out advanced exports on their own, I figure documentation is what this is for, right? Or if there's really nothing to cover that can be thought of, why is there an advanced section at all? At that point maybe rewriting that section would help to reduce confusion, as when I first came to this page early on, I know I tried looking for the advanced knowledge for 10-15 minutes before I realised it was never added to the page lol (I'm stupid and prob have memory loss bc that's happened multiple times, even later on LMAO)

It's been so long, I'm surprised it's still like that. I guess I half expected someone to point it out for me, but I don't think that's happening lol. I don't know why, it bugs me a lot

---
### 2.

Also, an aside (and the reason I came here and found out), there really should be a way to add more prefixes to `export_group`. I often have exported private variables, but if, for example, my prefix is `"navigation_"`, then `_navigation_abcd` would not be added to the group, as it begins with an underscore which is then followed by the prefix. 

Maybe there's a built-in way to do this already, but it was not made clear in the one section talking about `export_group`. So either mention that, or:

I suggest either allow alternative prefixes (and keep the first one as the group name) or simply ignore underscores at the start of variable names for this scenario. Also keep in mind with that decision, though, not everyone abides by or wants to abide by GDScript's official standards: Some people might use `m_variable`, `__variable`, `p_variable`, or others. I know if I still found real reason to mix C# and GDScript in my projects, I would much prefer to keep the naming of C# when using GDScript, and I personally would then be using `m_variable`. So really, it may just be better to add optional parameters to `export_group` to not have to worry about that. (Maybe the underscores could be auto, though.)

---
### 3.

This is not a request, this is a cry for help. And it's in no way related to exports.

PLEASE ADD ACCESS MODIFIERS AND OVERLOADING TO GDSCRIPT. OH MY GOD I HATE NOT HAVING THEM. THE ONLY REASON I USE GDSCRIPT IS FOR PERFORMANCE AND BREVITY. I DISLIKE SO MUCH ABOUT GDSCRIPT, BUT IT'S WITHOUT A DOUBT THE GENERALLY BEST OPTION. AND WHY IS THE EDITOR SO MID TO MAKE ME USE VSCODE'S SLIGHTLY BETTER ALTERNATIVE? AND WHY DOES EVERYTHING ERROR AND REQUIRE A RELOAD EVERY TWO SECONDS? MY BELOVED C# WAS SACRIFICED FOR THE GDSCRIPT TEAM'S SINS!!! 

thank you. and buff c# pls

i know where you live btw ;)

jk, but i will find out soon if this is not added >:(

---

Anyway yeah, thanks for reading lol

For legal reasons, I was joking about finding where you live...?

Anyway, back to coding! I hope this comment wasn't for nothing, but I doubt my GDScript requests will be done seeing how nothing has progressed in that direction in the whole of its existence lol. I use this engine for free, though, so I have no room to complain, just a genuine urge to lmao. Although complaints are good to know for software of any kind...so I spread the word anywhere I can lol

**Reply by Calinou** [reply]
Please open documentation improvement requests on https://github.com/godotengine/godot-docs/issues after searching for duplicates.

PS: I have to remind you that we have a [Code of Conduct](https://godotengine.org/code-of-conduct). Please stay constructive.

**Comment by tossudev** [comment]
Exported floating point values default their precision to 3 decimals at most. Is this intentional and if it is, should it be documented?

**Reply by Calinou** [reply]
This is intentional, and is configurable with the **Default Float Step** editor setting.

If you want consistent behavior regardless of editor settings, you should always define a step explicitly in the exported property hint. See discussion in https://github.com/godotengine/godot-proposals/issues/3524.

**Comment by Radivarig** [comment]
You can write `@export_tool_button` right above and skip one assignment if you define a function like this:
```gdscript
@export_tool_button("Print Yea")
var some_func_right_after: Callable = func():
	print("Yea")
```

Also, would be useful to mention that the scene needs to be reloaded to get rid of Nil error:
> The value of property "some_func_right_after" is Nil, but Callable was expected.

**Reply by Rouli-M** [reply]
YES! Thank you! How are we supposed to know that?

**Reply by Radivarig** [reply]
@Rouli-M I've edited above with a note

**Reply by Radivarig** [reply]
@Rouli-M nevermind, it's a 4.8-dev2 regression tracked here https://github.com/godotengine/godot/issues/121729

**Comment by lfod1997** [comment]
> To update it, call `notify_property_list_changed()` after setting the exported variable's value.

...and also call `EditorInterface.mark_scene_as_unsaved()` if the property of a node is INDEED changed (you may need to compare the new value with the old in your tool script, meaning elem-wise comparison if an `Array` or `Dictionary` is changed).

Btw, I'm finding it impossible to support undo-redo in such cases. I've read [EditorPlugin.get_undo_redo](https://docs.godotengine.org/en/stable/classes/class_editorplugin.html#class-editorplugin-method-get-undo-redo) but that one's limited to dedicated `EditorPlugin`s. Any light shed on this topic?

**Comment by Davolution** [comment]
Not sure if this has been discussed before, and I'm uncertain whether it adds any value, but I wanted to share something I discovered while experimenting with **@export_group**.

I found that using "/" in the group name forces the creation of all subgroup levels down to where the next export variable is located, whereas using @export_subgroup only displays the final menu level if no previous levels contain exported variables.

Thought this might be worth mentioning! 

ie: 
` @export_group("Loots / Drops / Items / Consumables / Potions / Magic / Attack / Colors")`
` @export var potion1 := "Attack up"`

Image:
<image>https://pasteboard.co/msxoPi4ZhPaJ.png

GDscript:
<image>https://pasteboard.co/LJ1ZoV6dZrb3.png

**Comment by XandaPanda42** [comment]
Could we include a note about the timeline for `@export`? It's not stated anywhere (that I can find at least) that the value set in the inspector isn't available during the Object's  `_init` method (often useful for custom Resources). Eg:

```GDScript
@export var exported_val: float

func _init() -> void:
    #exported_val isn't set yet and will be seen as null or zero depending on the type.
    print(exported_val) #Prints `0` and not whatever is set in the inspector.
```

Setting the value on the same line will allow the value to be used in `_init`, but the Inspector value still won't be used. Eg:

```GDScript
@export var exported_val: float = 1

func _init() -> void:
    print(exported_val) # Prints 1
```

A semi decent way around this is to use a setter to call a method, and avoid using `_init` for exported properties.

```GDScript
# Set to 3 in the inspector.
@export var exported_val: float = 2:
    set(val):
        exported_val = val
        print("Inspector Val: ", exported_val)

func _init() -> void:
    print("Init Val: ", exported_val)
```

Resulting in
> Init Val: 2
> Inspector Val: 3

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11308

**Comment by mcmodderHD** [comment]
In Godot 4.5 the following Code does no longer work

```gdscript 
@export_tool_button("Hello", "Callable") var hello_action = hello
```

The `Callable` has to be removed, it should look like this:

```gdscript 
@export_tool_button("Hello") var hello_action = hello
```

Edit: Okay nevermind I just had to reload the scene. But it still works without the Callabe.

**Reply by Radivarig** [reply]
It should either be fixed so no scene reload is needed or to document that it does.

**Comment by EloiStree** [comment]
If you have a nil from the export_tool_button... 


> @export_tool_button("Update Distance Compute Data","Callable")
> var some_func_right_after: Callable = func(): update_distance_compute_data()

Restart Godot. 
Worked for me after one hour of reading doc around.

**Comment by moonvent** [comment]
export_storage: Doesnt correct work. In my script I has all vars are export, and want to not show them in editor. Okay, I try to use export storage insteand and boom! It's doesnt work!!!! Not like export, not like all, I try to save / load my scene, and I get some phantom error. Them just doesn't work with that :( Return export and script work fine.

**Reply by Calinou** [reply]
Please don't ask support questions here, as this page is meant to be used for documentation notes only. Use one of the other [community channels](https://godotengine.org/community) instead.

**Reply by moonvent** [reply]
But this is not a support question, something in documentation wrong, because it doesn't work

**Comment by inces86** [comment]
I can't learn from this documentation how do I export String from predefined const array,like:
const animals = ["Horse","Cat","Dog"]
@export var chosen_animal ????

It is important that array is separate property, because others are supposed to refer to it
Don't tell me this is not a feature

**Comment by vladtepesch** [comment]
Is there a way to define aliases for annotations?
do define a shortcut for things like: `@export_range(0, 100, 1, "or_less", "or_greater")` if I have multiple variables that share some concept

**Reply by Calinou** [reply]
There currently isn't a way to define custom annotations, so this isn't feasible.

**Comment by meloonics** [comment]
while using `@export_range()` I got recommended the hint-string "prefer_slider"` in the code completion. I wanted to look up what precisely it means, but i could not find anything here, which to me seems like the most appropriate place to look.
According to a glance through the source it turns out it means, that a slider for integer values is used, instead of a spinbox. neato!

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11836

**Comment by p-planet** [comment]
curious if there is a way to export the lock/unlock component ratio button on Vector3s?

**Reply by meloonics** [reply]
`PropertyHint.PROPERTY_HINT_LINK`
Edit1: Oh apologies, I thought this worked for both vec2 and 3, sorry 😐
Edit2: oh, it does work in 4.6.1.stable, though the docs don't explicitly mention vector3s. 
But for what it's worth, paste this:
`@export_custom(PROPERTY_HINT_LINK, "") var my_vec3 : Vector3`

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11872
